activemq-broker/src/main/java/org/apache/activemq/replica/ReplicaCompactor.java

```diff
@@ -0,0 +1,303 @@
+import org.apache.activemq.command.MessageId;
+import org.apache.activemq.command.TransactionId;
+import org.apache.activemq.util.LongSequenceGenerator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ReplicaCompactor {
+    private static final Logger logger = LoggerFactory.getLogger(ReplicaCompactor.class);
+    private static final String CONSUMER_SELECTOR = String.format("%s LIKE '%s'", ReplicaEventType.EVENT_TYPE_PROPERTY, ReplicaEventType.MESSAGE_ACK);
+    public static final int MAXIMUM_MESSAGES = 1_000;
+
+    private final Broker broker;
+    private final ConnectionContext connectionContext;
+    private final ReplicaReplicationQueueSupplier queueProvider;
+    private final PrefetchSubscription subscription;
+
+    private final Queue intermediateQueue;
+
+    public ReplicaCompactor(Broker broker, ConnectionContext connectionContext, ReplicaReplicationQueueSupplier queueProvider, PrefetchSubscription subscription) {
+        this.broker = broker;
+        this.connectionContext = connectionContext;
+        this.queueProvider = queueProvider;
+        this.subscription = subscription;
+
+        intermediateQueue = broker.getDestinations(queueProvider.getIntermediateQueue()).stream().findFirst()
+                .map(DestinationExtractor::extractQueue).orElseThrow();
+    }
+
+    List<MessageReference> compactAndFilter(List<MessageReference> list, boolean withAdditionalMessages) throws Exception {
+        List<DeliveredMessageReference> toProcess = list.stream()
+                .map(DeliveredMessageReference::new)
+                .collect(Collectors.toList());
+
+        int prefetchSize = subscription.getPrefetchSize();
+        try {
+            if (withAdditionalMessages) {
+                subscription.setPrefetchSize(0);
+                toProcess.addAll(getAdditionalMessages());
+            }
+
+            List<DeliveredMessageReference> processed = compactAndFilter0(toProcess);
+
+            Set<MessageId> messageIds = list.stream().map(MessageReference::getMessageId).collect(Collectors.toSet());
+
+            return processed.stream()
+                    .map(dmr -> dmr.messageReference)
+                    .filter(mr -> messageIds.contains(mr.getMessageId()))
+                    .collect(Collectors.toList());
+        } finally {
+            subscription.setPrefetchSize(prefetchSize);
+        }
+    }
+
+    private List<DeliveredMessageReference> getAdditionalMessages() throws Exception {
+        List<DeliveredMessageReference> result = new ArrayList<>();
+        List<QueueMessageReference> additionalMessages = intermediateQueue.getMatchingMessages(connectionContext, CONSUMER_SELECTOR, MAXIMUM_MESSAGES);
```

Comment:
we expect to have only 2 messages here - one for source and one for replica. so it should impact the performance

Reply:
it will do a full scan only if there are less than 1000 messages for this selector. but yeah, I agree it may be expensive. I will take a look what we can do about it. but any ideas are highly appreciated